neps/nep-0509.md

```diff
@@ -0,0 +1,592 @@
+```
+
+The more stake validator has, the more **heavy** work it will get assigned, because we assume that validators with higher stakes have more powerful hardware. 
+With stateless validation, relative heaviness of the work changes. Comparing to the current order "block production" > "chunk production", the new order is "chunk production" > "block production" > "chunk validation".
+
+Shards are equally split among chunk producers: as in Mainnet on 12 Jun 2024 we have 6 shards, each shard would have ~16 chunk producers assigned.
+
+In the future, with increase in number of shards, we can generalise the assignment by saying that each shard should have `X` chunk producers assigned, if we have at least `X * S` validators. In such case, pseudocode for the role assignment would look as follows:
+
+```python
+if index(validator) < X * S:
+    roles(validator).append("chunk producer")
+if index(validator) < 100:
+    roles(validator).append("block producer")
+roles(validator).append("chunk validator")
+```
+
+#### Rewards
+
+Reward for each validator is defined as `total_epoch_reward * validator_relative_stake * work_quality_ratio`, where:
+
+* `total_epoch_reward` is selected so that total inflation of the token is 5% per annum;
+* `validator_relative_stake = validator_stake / total_epoch_stake`;
+* `work_quality_ratio` is the measure of the work quality from 0 to 1.
+
+So, the actual reward never exceeds total reward, and when everyone does perfect work, they are equal.
+For the context of the NEP, it is enough to assume that `work_quality_ratio = avg_{role}({role}_quality_ratio)`.
+So, if node is both a block and chunk producer, we compute quality for each role separately and then take average of them.
+  
+When epoch is finalized, all block headers in it uniquely determine who was expected to produce each block and chunk.
+Thus, if we define quality ratio for block producer as `produced_blocks/expected_blocks`, everyone is able to compute it.
+Similarly, `produced_chunks/expected_chunks` is a quality for chunk producer. 
+It is more accurate to say `included_chunks/expected_chunks`, because inclusion of chunk in block is a final decision of a block producer which defines success here.
+
+Ideally, we could compute quality for chunk validator as `produced_endorsements/expected_endorsements`. Unfortunately, we won't do it in Stage 0 because:
+
+* Mask of endorsements is not part of the block header, and it would be a significant change;
+* Block producer doesn't have to wait for all endorsements to be collected, so it could be unfair to say that endorsement was not produced if block producer just went ahead.
+
+So for now we decided to compute quality for chunk validator as ratio of `included_chunks/expected_chunks`, where we iterate over chunks which node was expected to validate.
+The obvious drawback here is that if chunks are not produced at all, chunk validators will also be impacted. We plan to address it in the future releases.
```

Comment:
A more important drawback of this approach is that chunk validators are not incentivized to validate the chunks, given they will be rewarded the same in either case.

Reply:
added